src/main/kotlin/com/salesforce/dynamodbv2/mt/sharedtable/MtBackupManager.kt

```diff
@@ -0,0 +1,84 @@
+/* Copyright (c) 2019, Salesforce.com, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause.
+ * For full license text, see LICENSE file in the repo root or https://opensource.org/licenses/BSD-3-Clause.
+ */
+package com.salesforce.dynamodbv2.mt.sharedtable
+
+import com.salesforce.dynamodbv2.mt.context.MtAmazonDynamoDbContextProvider
+import com.salesforce.dynamodbv2.mt.mappers.MtAmazonDynamoDbBase
+import com.salesforce.dynamodbv2.mt.mappers.sharedtable.impl.MtAmazonDynamoDbBySharedTable
+import com.salesforce.dynamodbv2.mt.sharedtable.impl.TenantTable
+
+/**
+ * Interface for grabbing backups of data managed by mt-dynamo.
+ *
+ * Backups are generated across all managed tables to individual tenant-table backups, which can independently
+ * be restored onto either the same environment, or migrated into different environments, with say, different
+ * multi-tenant strategies (ie: moving a tenant-table from a table per tenant setup onto a shared table setup, or
+ * vice versa.
+ *
+ * One more dimension of value added by these backups are, they should be mt-dynamo version agnostic. So if a backup
+ * was generated at v0.10.5 of mt-dynamo, and imagine the physical representation of tenant to table mapping strategy
+ * changes at v0.11.0, that backup should be restorable at that later version, offering a path to preserve data when
+ * we change table mappings as mt-dynamo evolves.
+ *
+ * At the moment, these backups are taking full snapshots of an mt-dynamo account, but there are plans to support PITR
+ * style continuous backups, offering a time window of available restore points (versus choosing from N snapshots)
+ */
+interface MtBackupManager {
+    /**
+     * Go through all tenants owned by this instance, and create a backup snapshot on S3 per table-tenant.
+     */
+    fun createMtBackup(createMtBackupRequest: CreateMtBackupRequest, mtDynamo: MtAmazonDynamoDbBySharedTable): MtBackupMetadata
+
+    /**
+     * Get the status of a given multi-tenant backup.
+     */
+    fun getBackup(id: String): MtBackupMetadata?
+
+    /**
+     * Delete the give multi-tenant backup.
+     */
+    fun deleteBackup(id: String): MtBackupMetadata?
+
+    /**
+     * Get details of a given table-tenant backup.
+     */
+    fun getTenantTableBackup(id: String): TenantTableBackupMetadata
+
+    /**
+     * Initiate a restore of a given table-tenant backup to a new table-tenant target.
+     */
+    fun restoreTenantTableBackup(
+        restoreMtBackupRequest: RestoreMtBackupRequest,
+        mtDynamo: MtAmazonDynamoDbBase,
+        mtContext: MtAmazonDynamoDbContextProvider
+    ): TenantRestoreMetadata
+
+    /**
+     * List all multi-tenant backups known to us on S3.
+     */
+    fun listMtBackups(): List<MtBackupMetadata>
+}
+
+data class MtBackupMetadata(val mtBackupId: String, val status: Status, val tenantTables: Set<TenantTableBackupMetadata>)
```

Comment:
being able to define data classes like this without needing lombok is maybe one of my favorite parts of Kotlin. So much support for keeping code concise, with extensions for customizing equals or toString() or anything else you'd want to customize.